3MI0800270/2023-02-28/zad1.cpp

```diff
@@ -0,0 +1,80 @@
+#include <iostream>
+using namespace std;
+
+struct transition {
+    size_t to;
+    char c;
+};
+
+struct state {
+    transition transitions[4];
+    size_t num;
+};
+
+struct stateInfo {
+    state data[50];
+};
+
+bool hasTransition(const stateInfo& arg, size_t from, char c) {
+    size_t number = arg.data[from].num;
+    for (int j = 0;j < number;j++) {
+        if (arg.data[from].transitions[j].c == c) {
+            return true;
+        }
+    }
+}
+bool addTransition(stateInfo& arg, size_t from, char c, size_t to) {
+    size_t number = arg.data[from].num;
+    if (number >= 4) {
+        return false;
+    }
+    else {
+        arg.data[from].transitions[number].c = c;
+        arg.data[from].transitions[number].to = to;
+    }
+}
+int getTransition(const stateInfo& arg, size_t from, char c) {
+    size_t number = arg.data[from].num;
+    for (int i = 0;i < number;i) {
+        if (arg.data[from].transitions[i].c == c) {
+            return i;
+        }
+    }
+    return -1;
+}
+bool canRead(const stateInfo& arg, char* word, size_t from) {
+    size_t pos = from;
+    while (!*word) {
+        if (hasTransition(arg, pos, *word)) {
+            pos = arg.data[pos].transitions[getTransition(arg, pos, *word)].to;
+            word++;
+        }
+        else {
+            break;
+        }
+    }
+    if (*word) {
+        return true;
+    }
+    else {
+        return false;
+    }
+}
+bool removeTransition(stateInfo& arg, size_t from, char c) {
```

Comment:
It's better and cleaner to write this like so:

```
if (!hasTransition(arg, from, c)) {
    return false;
}

size_t pos = getTransition(arg, from, c);
size_t number = arg.data[from].num;
arg.data[from].transitions[pos].c = arg.data[from].transitions[number].c;
arg.data[from].transitions[pos].to = arg.data[from].transitions[number].to;
arg.data[from].num--;
return true;
```

This makes life easier for whoever is reading the code - the 'corner' case has been visually outlined and dealt with in the beginning of the function.

Reply:
Just out of curiosity - isn't `num` containing the total number of transitions ? If this is the case - `number` will always be bigger than the last index of an element in the `transitions` array :)) In addition to that - instead of the 'manual' swap you're doing (lines 67 & 68) you can use [std::swap](https://en.cppreference.com/w/cpp/algorithm/swap) directly :thinking: